### PR TITLE
llvmPackages_11: start with 11.0.0rc2

### PR DESCRIPTION
**Pull request by ggreif**
http://lists.llvm.org/pipermail/release-testers/2020-August/001321.html
http://lists.llvm.org/pipermail/release-testers/2020-July/001305.html

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by conferno** [comment]
I tried to build and it fails
```
 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: DeLICM/./DeLICMTests/failed_to_discover_tests_from_gtest (39519 of 39568)
******************** TEST 'Polly-Unit :: DeLICM/./DeLICMTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-delicm-max-ops' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: DeLICM/./DeLICMTests/failed_to_discover_tests_from_gtest (39520 of 39568)
******************** TEST 'Polly-Unit :: DeLICM/./DeLICMTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-delicm-max-ops' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

 99% [==========================================================-(B](B ETA: 00:00:00
 99% [==========================================================-(B](B ETA: 00:00:00
 99% [==========================================================-(B](B ETA: 00:00:00
 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: ScheduleOptimizer/./ScheduleOptimizerTests/failed_to_discover_tests_from_gtest (39524 of 39568)
******************** TEST 'Polly-Unit :: ScheduleOptimizer/./ScheduleOptimizerTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-opt-optimize-only' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

 99% [==========================================================-(B](B ETA: 00:00:00
 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: ScopPassManager/./ScopPassManagerTests/failed_to_discover_tests_from_gtest (39526 of 39568)
******************** TEST 'Polly-Unit :: ScopPassManager/./ScopPassManagerTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-detect-profitability-min-per-loop-insts' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: ScheduleOptimizer/./ScheduleOptimizerTests/failed_to_discover_tests_from_gtest (39527 of 39568)
******************** TEST 'Polly-Unit :: ScheduleOptimizer/./ScheduleOptimizerTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-opt-optimize-only' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

 99% [==========================================================-(B](B ETA: 00:00:00
FAIL: Polly-Unit :: ScopPassManager/./ScopPassManagerTests/failed_to_discover_tests_from_gtest (39528 of 39568)
******************** TEST 'Polly-Unit :: ScopPassManager/./ScopPassManagerTests/failed_to_discover_tests_from_gtest' FAILED ********************
: CommandLine Error: Option 'polly-detect-profitability-min-per-loop-insts' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

********************
                     -- Testing: 39568 tests, 12 workers --                     (B

```

**Comment by conferno** [comment]
@GrahamcOfBorg build llvm_11

**Comment by conferno** [comment]
@GrahamcOfBorg build clang_11

**Comment by conferno** [comment]
@GrahamcOfBorg build lld_11

File: pkgs/development/compilers/llvm/11/llvm.nix [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+, zlib
+, buildPackages
+, debugVersion ? false
+, enableManpages ? false
+, enableSharedLibraries ? true
+, enablePFM ? !(stdenv.isDarwin
+  || stdenv.isAarch64 # broken for Ampere eMAG 8180 (c2.large.arm on Packet) #56245
+  || stdenv.isAarch32 # broken for the armv7l builder
+)
+, enablePolly ? true
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/llvm.nix:21]:
```suggestion
, enablePolly ? false
```

this does fix the build (it is `false` in all `llvm_*` except `10`)

File: pkgs/development/compilers/llvm/11/compiler-rt.nix [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    "-DCMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY"
+  ] ++ stdenv.lib.optionals (bareMetal) [
+    "-DCOMPILER_RT_OS_DIR=baremetal"
+  ];
+
+  outputs = [ "out" "dev" ];
+
+  patches = [
+    ./compiler-rt-codesign.patch # Revert compiler-rt commit that makes codesign mandatory
+    ./find-darwin-sdk-version.patch # don't test for macOS being >= 10.15
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/compiler-rt.nix:50]:
```suggestion
```
the patch seems to included upstream and not needed

File: pkgs/development/compilers/llvm/11/clang/purity.patch [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+       CmdArgs.push_back("-export-dynamic");
+-
+-    if (!Args.hasArg(options::OPT_shared) && !IsStaticPIE) {
+-      const std::string Loader =
+-          D.DyldPrefix + ToolChain.getDynamicLinker(Args);
+-      CmdArgs.push_back("-dynamic-linker");
+-      CmdArgs.push_back(Args.MakeArgString(Loader));
+-    }
+   }
+
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/clang/purity.patch:26]: this patch is from llvm_10, it fails to apply to llvm_11 and need to be altered

```diff
 index fe3c0191bb..c6a482bece 100644
 --- a/lib/Driver/ToolChains/Gnu.cpp
 +++ b/lib/Driver/ToolChains/Gnu.cpp
-@@ -494,13 +494,6 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+@@ -487,12 +487,6 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
    if (!IsStatic) {
      if (Args.hasArg(options::OPT_rdynamic))
        CmdArgs.push_back("-export-dynamic");
 -
 -    if (!Args.hasArg(options::OPT_shared) && !IsStaticPIE) {
--      const std::string Loader =
--          D.DyldPrefix + ToolChain.getDynamicLinker(Args);
 -      CmdArgs.push_back("-dynamic-linker");
--      CmdArgs.push_back(Args.MakeArgString(Loader));
+-      CmdArgs.push_back(Args.MakeArgString(Twine(D.DyldPrefix) +
+-                                           ToolChain.getDynamicLinker(Args)));
 -    }
    }
```

File: pkgs/development/compilers/llvm/11/clang/default.nix [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      "-DLLVM_ENABLE_SPHINX=ON"
+      "-DSPHINX_OUTPUT_MAN=ON"
+      "-DSPHINX_OUTPUT_HTML=OFF"
+      "-DSPHINX_WARNINGS_AS_ERRORS=OFF"
+    ];
+
+    patches = [
+      ./purity.patch
+      # https://reviews.llvm.org/D51899
+      ./compiler-rt-baremetal.patch
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/clang/default.nix:41]:
```suggestion
```

no need, merged upstream

**Comment by conferno** [comment]
With the fixes above, I managed to build `llvm 11` and `chromium-git 86` using it

**Comment by ggreif** [comment]
Yes, removing `.patch` files next.

File: pkgs/development/compilers/llvm/11/clang/purity.patch (hunk 3 above)
Comment (ggreif) [reply pkgs/development/compilers/llvm/11/clang/purity.patch:26]: Feel free to make a multi-line suggestion!

File: pkgs/development/compilers/llvm/11/clang/purity.patch [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+-          D.DyldPrefix + ToolChain.getDynamicLinker(Args);
+-      CmdArgs.push_back("-dynamic-linker");
+-      CmdArgs.push_back(Args.MakeArgString(Loader));
+-    }
+   }
+ 
+   CmdArgs.push_back("-o");
+-- 
+2.11.0
+
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/clang/purity.patch:30]:
```suggestion
From 4add81bba40dcec62c4ea4481be8e35ac53e89d8 Mon Sep 17 00:00:00 2001
From: Will Dietz <w@wdtz.org>
Date: Thu, 18 May 2017 11:56:12 -0500
Subject: [PATCH] "purity" patch for 5.0

---
 lib/Driver/ToolChains/Gnu.cpp | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/lib/Driver/ToolChains/Gnu.cpp b/lib/Driver/ToolChains/Gnu.cpp
index fe3c0191bb..c6a482bece 100644
--- a/lib/Driver/ToolChains/Gnu.cpp
+++ b/lib/Driver/ToolChains/Gnu.cpp
@@ -487,12 +487,6 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   if (!IsStatic) {
     if (Args.hasArg(options::OPT_rdynamic))
       CmdArgs.push_back("-export-dynamic");
-
-    if (!Args.hasArg(options::OPT_shared) && !IsStaticPIE) {
-      CmdArgs.push_back("-dynamic-linker");
-      CmdArgs.push_back(Args.MakeArgString(Twine(D.DyldPrefix) +
-                                           ToolChain.getDynamicLinker(Args)));
-    }
   }
 
   CmdArgs.push_back("-o");
-- 
2.11.0

```

File: pkgs/development/compilers/llvm/11/clang/purity.patch [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+     if (Args.hasArg(options::OPT_rdynamic))
+       CmdArgs.push_back("-export-dynamic");
+-
+-    if (!Args.hasArg(options::OPT_shared) && !IsStaticPIE) {
+-      const std::string Loader =
+-          D.DyldPrefix + ToolChain.getDynamicLinker(Args);
+-      CmdArgs.push_back("-dynamic-linker");
+-      CmdArgs.push_back(Args.MakeArgString(Loader));
+-    }
+   }
```
Comment (conferno) [inline pkgs/development/compilers/llvm/11/clang/purity.patch:25]:
```suggestion
@@ -487,12 +487,6 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   if (!IsStatic) {
     if (Args.hasArg(options::OPT_rdynamic))
       CmdArgs.push_back("-export-dynamic");
-
-    if (!Args.hasArg(options::OPT_shared) && !IsStaticPIE) {
-      CmdArgs.push_back("-dynamic-linker");
-      CmdArgs.push_back(Args.MakeArgString(Twine(D.DyldPrefix) +
-                                           ToolChain.getDynamicLinker(Args)));
-    }
   }
```

File: pkgs/development/compilers/llvm/11/clang/purity.patch (hunk 3 above)
Comment (conferno) [reply pkgs/development/compilers/llvm/11/clang/purity.patch:25]: done

Comment (ggreif) [reply pkgs/development/compilers/llvm/11/clang/purity.patch:25]: Oopps, I committed too soon. Should I touch it up?

**Comment by conferno** [comment]
@GrahamcOfBorg build clang_11

**Comment by ggreif** [comment]
I get
```
CMake Error at CMakeLists.txt:4 (message):
  libc++abi now requires being built in a monorepo layout with libcxx
  available
```

**Comment by conferno** [comment]
> I get
> 
> ```
> CMake Error at CMakeLists.txt:4 (message):
>   libc++abi now requires being built in a monorepo layout with libcxx
>   available
> ```

ofborg built is successfully

File: pkgs/development/compilers/llvm/11/clang/purity.patch (hunk 3 above)
Comment (conferno) [reply pkgs/development/compilers/llvm/11/clang/purity.patch:25]: No need, they are the same.
The first `suggestion` was a whole file diff, while the second is limited to the changed part.

**Comment by conferno** [comment]
What left is to fix build when `enablePolly = true`

**Comment by primeos** [comment]
Any news on this? Would be good to have this soon for testing `chromiumBeta` (though I could also apply this PR locally).

Is `enablePolly = true` the only problem that's left?

Also cc @lovek323 @7c6f434c @dtzWill (LLVM maintainers)

**Comment by ggreif** [comment]
Unfortunately the macOS build fails with https://github.com/NixOS/nixpkgs/pull/94204#issuecomment-667957332 and. I have no time right now to fix it. Probably easy to build a monorepo-like layout frame the downloaded tarballs, and I assume that not even all are needed. But I have no time right now to do the gruntwork, so we can either merge this and accept that Darwin is broken, or start working on it.

**Comment by primeos** [comment]
Thanks for the update. Given that this is a new package I would be fine with getting the Darwin support in place later if we have to.

@NixOS/darwin-maintainers do you maybe have time to take a look at the Darwin build?

**Comment by primeos** [comment]
Ok, so I just refactored/squashed the history and merged this in 28aef45a1aaf9b74ca2f5f1d015b687fee92a080 :)

Thanks for your work @ggreif and @conferno :heart:

**Pull request closed without merging**